UBCFormulaElectric/Consolidated-Firmware
# WIP: Refactor PDM

### Summary

WARNING: Don't merge. This is another big PR i'm trying to break down into smaller PRs.

### Changelist 

- `PDM.uvprojx`: Re-arrange sources files/headers in alphabetical order, Update header include path
- `PDM.uvoptx`: Add .c and .h files
- Remove `converted_reading[]` (Refactored into `profet[i].efuse[j].current` and `voltage_sense[]`) and `e_fuse_fault_states`  (Refactored into `profet[i].efuse[j].current_num_of_fault` and profet[i].efuse[j].max_num_of_fault` 
- Removed the heartbeat code in `stm32f3xx_it.c`. This is so we can use Kiron's new heartbeat library. See: #288.
- Introduced `static Profet2_Struct profet2s[NUM_PROFET2S]`, which greatly simplified a ton of helper functions related to Profet2 (e.g. Toggling DEN pin for each Profet2)
- Enabled `ADC_CHANNEL_VREFINT` in STM32CubeMX
- Use `USER_CONSTANTS` for watchdog related parameters in STM32CubeMX
- Add a bunch of `enums` in `Profet2.h` to make helper functions more clear
- Refactor the timer library by using `#define` to rename timer handles (e.g. `TIM2` -> `DMA_TIMER`)

### Testing Done

- Manually tested each of the new helper functions that use `static Profet2_Struct profet2s[NUM_PROFET2S]` in `Gpio.c` and `Profet2.c`
- Test that `CurrentSense_TransmitCurrent()` works on all GLV outputs
    - Start the log trace in PCAN Explorer 6 
    - For each output, disconnect the load so it's 0A for 10 seconds. Then, connect a 4Ohm resistors (Use the 50W resistors - they are yellow and chassis mount) for another 10 seconds
    - Repeat the above step for all outputs
    - Plot the log trace in Excel and make sure the values are reasonable (Upload the log to this PR)
- Test `VoltageSense_TransmitVoltage()` works on all voltage sense:
    - Just make sure that the CAN messages are reporting sensible values for VBAT (~8.4V), GLV (~12V), FLYWIRE (12V if the external 12V input is connected, or 0V if the external 12V input is disconnected)

### Resolved Issues


### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code standards detailed in http://svformulaep1.teams.apsc.ubc.ca:8090/display/TR/Workflow (*This will save time for both you and the reviewer!*).
